### ISSUE TITLE
AZ_SPAN confusion - how to set a device id programatically?

### ISSUE THREAD

**Issue by torntrousers** [Azure/azure-sdk-for-c#2084]
Hi, I can't figure out how to change the [azure-sdk-for-c-arduino](https://github.com/Azure/azure-sdk-for-c-arduino/tree/main/examples/Azure_IoT_Central_ESP32_AzureIoTKit) sample so that the device id can be set programatically. In the current sample its set with a #define [here](https://github.com/Azure/azure-sdk-for-c-arduino/blob/main/examples/Azure_IoT_Central_ESP32_AzureIoTKit/iot_configs.h#L10), and then used with an AZ_SPAN_FROM_STR [here](https://github.com/Azure/azure-sdk-for-c-arduino/blob/main/examples/Azure_IoT_Central_ESP32_AzureIoTKit/Azure_IoT_Central_ESP32_AzureIoTKit.ino#L341). I want to be able to update that line so that the device id is taken from a variable, eg: `char myDevID[65];` but  whatever i try gives a compile error, eg:
```
In file included from C:\Arduino\arduino-1.8.13\portable\sketchbook\libraries\Azure_SDK_for_C\src/az_base64.h:20,
                 from C:\Arduino\arduino-1.8.13\portable\sketchbook\libraries\Azure_SDK_for_C\src/az_core.h:18,
                 from C:\hms\azure-sdk-for-c-arduino\examples\Azure_IoT_Central_ESP32\Azure_IoT_Central_ESP32.ino:48:
C:\hms\azure-sdk-for-c-arduino\examples\Azure_IoT_Central_ESP32\Azure_IoT_Central_ESP32.ino: In function 'void setup()':
C:\Arduino\arduino-1.8.13\portable\sketchbook\libraries\Azure_SDK_for_C\src/az_span.h:125:50: error: expected primary-expression before ')' token
 #define AZ_SPAN_FROM_STR(STRING_LITERAL) (az_span) AZ_SPAN_LITERAL_FROM_STR(STRING_LITERAL)
                                                  ^
C:\hms\azure-sdk-for-c-arduino\examples\Azure_IoT_Central_ESP32\Azure_IoT_Central_ESP32.ino:357:42: note: in expansion of macro 'AZ_SPAN_FROM_STR'
   azure_iot_config.dps_registration_id = AZ_SPAN_FROM_STR(myDevID); // Use my deviceID
                                          ^~~~~~~~~~~~~~~~
```
How can i do this?

TIA!

**Comment by danewalton** [comment]
Hi @torntrousers 

Here we have an API to create an `az_span` which can be used to create a span over a dynamically created string ([link here](https://github.com/Azure/azure-sdk-for-c/blob/973a728cd590d1bcca4cd7632a46d151ba216c46/sdk/inc/azure/core/az_span.h#L59-L76)).

The API `AZ_SPAN_FROM_STR(myDevID)` should only be used for string literals (known at compile time).

**Comment by torntrousers** [comment]
Thanks @danewalton  but I still can't figure out what I should do. If i try this:
```
  char myDevID[65] = "1df33b6ac17a2273d693c4c234d89debffcdf73c2511c180d2844398100c0b75";
  azure_iot_config.dps_registration_id = AZ_NODISCARD(myDevID, 65);
```
i get this error:
```
n file included from C:\Arduino\arduino-1.8.13\portable\sketchbook\libraries\Azure_SDK_for_C\src/_az_cfg_prefix.h:29,
                 from C:\Arduino\arduino-1.8.13\portable\sketchbook\libraries\Azure_SDK_for_C\src/az_result.h:21,
                 from C:\Arduino\arduino-1.8.13\portable\sketchbook\libraries\Azure_SDK_for_C\src/az_base64.h:19,
                 from C:\Arduino\arduino-1.8.13\portable\sketchbook\libraries\Azure_SDK_for_C\src/az_core.h:18,
                 from C:\hms\azure-sdk-for-c-arduino\examples\Azure_IoT_Central_ESP32\Azure_IoT_Central_ESP32.ino:48:
C:\hms\azure-sdk-for-c-arduino\examples\Azure_IoT_Central_ESP32\Azure_IoT_Central_ESP32.ino: In function 'void setup()':
C:\Arduino\arduino-1.8.13\portable\sketchbook\libraries\Azure_SDK_for_C\src/_az_cfg.h:83:22: error: expected primary-expression before '__attribute__'
 #define AZ_NODISCARD __attribute__((warn_unused_result))
                      ^~~~~~~~~~~~~
C:\hms\azure-sdk-for-c-arduino\examples\Azure_IoT_Central_ESP32\Azure_IoT_Central_ESP32.ino:358:42: note: in expansion of macro 'AZ_NODISCARD'
   azure_iot_config.dps_registration_id = AZ_NODISCARD(myDevID, 65);
                                          ^~~~~~~~~~~~
```

**Comment by torntrousers** [comment]
Thanks, using `az_span_create` does fix it.
